bson/src/main/org/bson/internal/BsonUtil.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bson.internal;
+
+import org.bson.BsonArray;
+import org.bson.BsonBinary;
+import org.bson.BsonDocument;
+import org.bson.BsonJavaScriptWithScope;
+import org.bson.BsonValue;
+
+/**
+ * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ */
+public final class BsonUtil {
+    public static BsonDocument mutableDeepCopy(final BsonDocument original) {
+        BsonDocument copy = new BsonDocument(original.size());
+        original.forEach((key, value) -> copy.put(key, mutableDeepCopy(value)));
+        return copy;
+    }
+
+    private static BsonArray mutableDeepCopy(final BsonArray original) {
+        BsonArray copy = new BsonArray(original.size());
+        original.forEach(element -> copy.add(mutableDeepCopy(element)));
+        return copy;
+    }
+
+    private static BsonBinary mutableDeepCopy(final BsonBinary original) {
+        return new BsonBinary(original.getType(), original.getData().clone());
+    }
+
+    private static BsonJavaScriptWithScope mutableDeepCopy(final BsonJavaScriptWithScope original) {
+        return new BsonJavaScriptWithScope(original.getCode(), mutableDeepCopy(original.getScope()));
+    }
+
+    private static <T extends BsonValue> T mutableDeepCopy(final T original) {
```

Comment:
Since we have to suppress an unchecked warning anyway, is it any better/safer to use `T extends BsonValue ` here rather than just `BsonValue`?  I guess it serves as documentation of the intention even if it's not enforced by the compiler?

Reply:
This is bad. I wrote it this way to avoid casting everywhere I call it. But now that you asked, I checked, and everywhere the method is called (just two places), `BsonValue` is expected, so no, there is no reason for this method to be generic or to have any explicit conversions.